optimum/tpu/modeling_llama.py

```diff
@@ -340,6 +340,8 @@ def _init_rope(self):
                 base=self.rope_theta,
             )
         else:
+            # Set Default rope_scaling to dynamic
+            self.config.rope_scaling.setdefault("type", "dynamic")
```

Comment:
@tengomucho For Llama 3.1 I assume we need to replace `rope_scaling["type"]` to `rope_scaling['rope_type']`, which means for Llama 3.1 `scaling_type` would be `rope_scaling['rope_type']`. If this is the case should we create a new conditional branch for `scaling_type = "llama3"`? Below is the `rope_scaling` dict from `config.json` of Llama 3.1 with rope_type value "llama3"

```
"rope_scaling": {
    "factor": 8.0,
    "low_freq_factor": 1.0,
    "high_freq_factor": 4.0,
    "original_max_position_embeddings": 8192,
    "rope_type": "llama3"
  }
```
My draft implementation of _init_rope is as below: Please suggest

```
def _init_rope(self):
        if self.config.rope_scaling is None:
            self.rotary_emb = LlamaRotaryEmbedding(
                self.head_dim,
                max_position_embeddings=self.max_position_embeddings,
                base=self.rope_theta,
            )
        else:
            scaling_type = self.config.rope_scaling.get("rope_type", self.config.rope_scaling.get("type"))
            scaling_factor = self.config.rope_scaling["factor"]
            if scaling_type == "linear":
                self.rotary_emb = LlamaLinearScalingRotaryEmbedding(
                    self.head_dim,
                    max_position_embeddings=self.max_position_embeddings,
                    scaling_factor=scaling_factor,
                    base=self.rope_theta,
                )
            elif scaling_type == "dynamic":
                self.rotary_emb = LlamaDynamicNTKScalingRotaryEmbedding(
                    self.head_dim,
                    max_position_embeddings=self.max_position_embeddings,
                    scaling_factor=scaling_factor,
                    base=self.rope_theta,
                )
            elif scaling_type == "llama3"
            # Your suggestion required.
            ......
            ......
            ......
            else:
                raise ValueError(f"Unknown RoPE scaling type {scaling_type}")
```

Reply:
We could do that @Bihan , but that would mean that we would end up with a code that diverges more from the original transformers' [code](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py). I was suggesting to stay as close as possible to their implementation to simplify maintenance: if there is a new update to transformers to support a new feature or fix a bug, it will be easier to support it if optimum-tpu code is more similar.